File: CousinsInBT.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

class Solution {
    TreeNode parent_x;
    TreeNode parent_y;
    int depth_x;
    int depth_y;
    public boolean isCousins(TreeNode root, int x, int y) {
        dfs(root, null, 0, x, y);
        return parent_x!=parent_y && depth_x==depth_y;
    }
    public void dfs(TreeNode root, TreeNode parent, int depth, int x, int y){
        if(root==null) return;
        if(root.val == x) {
            parent_x=parent;
            depth_x=depth;
        }
        if(root.val == y) {
            parent_y=parent;
            depth_y=depth;
        }
        if(parent_x==null || parent_y==null) dfs(root.left, root, depth+1, x, y);
        if(parent_x==null || parent_y==null) dfs(root.right, root, depth+1, x, y);
    }
}

File: cousins in BT.py
# // Time Complexity : O(n)
# // Space Complexity : O(n)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No

class Solution:
    def isCousins(self, root, x: int, y: int) -> bool:
        
        # to track parent and depth
        tracker = {}
        def dfs(node,parent,depth,x,y):
            if not node:
                return 
            
            dfs(node.left,node,depth+1,x,y)
            dfs(node.right,node,depth+1,x,y)
            
            # add a tuple of parent and depth as value to the nodes (key) 
            if node.val in (x,y):
                tracker[node.val] = (parent,depth)
                
        dfs(root,None,0,x,y)
        
        # logic to check for cousins
        cousins = tracker.get(x) and tracker.get(y) and tracker[x][1] == tracker[y][1] and tracker[x][0] != tracker[y][0]
        
        return cousins